src/Database/LSMTree/Internal/Run/Construction.hs

```diff
@@ -0,0 +1,493 @@
+{-# LANGUAGE BangPatterns    #-}
+{-# LANGUAGE DeriveFoldable  #-}
+{-# LANGUAGE DeriveFunctor   #-}
+{-# LANGUAGE MagicHash       #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ViewPatterns    #-}
+{- HLINT ignore "Redundant lambda" -}
+{- HLINT ignore "Use camelCase" -}
+
+-- | Incremental, in-memory run consruction
+--
+module Database.LSMTree.Internal.Run.Construction (
+    -- * Types
+    KOp
+  , RawKey
+  , RawValue (RawValue, EmptyRawValue)
+  , Operation (..)
+  , BlobRef (..)
+    -- * Pages
+    -- ** Page Builder
+  , SizedBuilder (..)
+  , pageBuilder
+    -- **  Page accumulator
+  , PageAcc
+  , paIsEmpty
+  , paIsOverfull
+  , paEmpty
+  , paAddElem
+  , paSingleton
+    -- * Exposed only for testing
+    -- ** StricterList
+  , StricterList (..)
+    -- ** BitMap
+  , BitMap (..)
+  , Bit
+  , emptyBitMap
+  , appendBit
+  , boolToBit
+    -- ** CrumbMap
+  , CrumbMap (..)
+  , Crumb
+  , emptyCrumbMap
+  , appendCrumb
+  , boolsToCrumb
+  ) where
+
+import           Control.Exception (assert)
+import           Data.Bits
+import qualified Data.ByteString.Builder as BB
+import qualified Data.ByteString.Builder.Internal as BB
+import           Data.ByteString.Short.Internal (ShortByteString (SBS))
+import qualified Data.ByteString.Short.Internal as SBS
+import           Data.Maybe (fromMaybe, isJust)
+import           Data.Monoid (Dual (..))
+import           Data.Primitive.ByteArray
+import           Data.Word
+import qualified Database.LSMTree.Internal.Serialise as S
+import           Foreign.Ptr
+
+{-------------------------------------------------------------------------------
+  Types
+-------------------------------------------------------------------------------}
+
+-- | A k\/op pair with an optional blob reference.
+type KOp = (RawKey, Operation, Maybe BlobRef)
+
+kopKey :: KOp -> RawKey
+kopKey (x,_,_) = x
+
+kopOp :: KOp -> Operation
+kopOp (_,x,_) = x
+
+kopBlobRef :: KOp -> Maybe BlobRef
+kopBlobRef (_,_,x) = x
+
+--
+-- Keys
+--
+
+-- | A string of raw bytes representing a key.
+type RawKey = S.SerialisedKey
+
+-- | Size of key in number of bytes.
+sizeofKey :: RawKey -> Int
+sizeofKey (S.SerialisedKey ba) = sizeofByteArray ba
+
+-- | Size of key in number of bytes.
+sizeofKey16 :: RawKey -> Word16
+sizeofKey16 = fromIntegral . sizeofKey
+
+-- | Size of key in number of bytes.
+sizeofKey64 :: RawKey -> Word64
+sizeofKey64 = fromIntegral . sizeofKey
```

Comment:
I feel like those are general enough to go into the `Serialise` module, together with other functions that look at the internal representation, e.g. `fromShortByteString :: ShortByteString -> SerialisedKey` and `toShortByteString :: SerialisedKey -> ShortByteString` instead of matching on the constructors here in `toKey` and `rawKey`.

Also, what is the advantage of the `RawKey` synonym, just saving a few characters?

Reply:
I see `SerialisedKey` as a placeholder type until we commit to a representation for raw keys/values/blobs -- the type synonym is just a minor convenience to swap out the raw key type